Allow users to use GLOB_MARK flag with glob()

telemachus opened the pull request:
Most of the optional flags for glob wouldn't make much sense for Lua,
but GLOB_MARK is very useful and would require a lot of work for a user
of luaposix to add after the fact. (When GLOB_MARK is added to a call of
glob, directories are automatically given a final '/' in the list of
return values.)

The function signature changes from glob([pattern]) to glob([pattern],
[glob_mark]). Any truthy value passed as a second argument turns on
GLOB_MARK. The second argument is optional, and the flag defaults to
false. So users who don't want this extra can simply ignore it, and they
can contingue to call the function as glob() and get the shortcut for
the "*" pattern.

I haven't added tests yet. If there's interest in this patch, I'll spend
a little time learning specl and do so. (Looks like tests for glob
should go in spects/posix_spec.yaml, yes?)

gvvaughan commented:
Hi Peter,

Sure, that seems like a useful addition and is POSIX conforming.  I'd be happy to accept a complete patch with tests.

Now that I've fixed Travis CI at the HEAD of luaposix, you needn't install the full Specl stack; once you rebase your RP Travis will run your additional Specl checks and link them to the PR page at github.  All you need to do is follow the format of the existing spec YAML files, which is pretty intuitive, I think.

Please feel free to email me or ask here if there's anything I can do to help!

telemachus commented:
@gvvaughan Thanks. Here's what I've done:

+ I updated my changes to be more in sync with the coding style in the project.
+ I updated the Ldoc material, so that the new option to `glob` is documented. 
+ I added tests to posix_spec.yaml.

Please let me know if you want me to make any further changes.

gvvaughan commented:
Hey, awesome!  Thanks for working on this :)

The only thing that is incongruous compared with the rest of luaposix is not following the POSIX API in your implementation.  See https://github.com/luaposix/luaposix/blob/master/ext/posix/stdio.c for an example of exporting C constants out of posix.so; and then, if it still seems like a huge usability win, you could add a more Lua-ish caller to https://github.com/luaposix/luaposix/blob/master/lib/posix.lua.in.

telemachus commented:
I see what you mean; sorry for missing that. Let me just be clear that I understand what this might look like:

+ Export the C constant `GLOB_MARK` in `ext/posix/glob.c`.
+ Write a wrapper function that allows users to call `glob` in a more Lua-ish fasion in `lib/posix.lua.in`.

If that sounds right, one further question: Do you agree that the function call I created initially is reasonable? My thought was that any truthy value passed as the glob-mark parameter would turn on final slashes.

Thanks for working with me on this.

gvvaughan commented:
Yep, your bullet points are exactly what I was describing.

On the surface your original proposal seems okay, but I'm not a big fan of naked booleans in an API that might change, because we might find a good use for one or more of the other flags from this or a future POSIX spec and want to add them in too eventually, but then we have to remember what position bool arg corresponds to what flag.  Conversely: bitwise ORing exported C constants is ugly in Lua, so we should try to follow the lead of existing calls that we wrap for similar reasons.

A good compromise could be to make the new second argument to the Lua wrapper API be a string. In which case we should make passing anything other than `nil` or `"MARK"` raise argument an error; in future if we need to expand, we can add support for further strings naming other flags in that or subsequent arguments...

Alternatively, if you prefer, you could pass in a Lua Set (a table with keys holding member names, and their values always set to `true`) so that we don't tie up all the remaining argument positions forever (short of breaking backwards compatibility).  This would make the implementation faster (just check `(flags or {}).MARK == true`), but calling would be a bit more verbose (i.e. `posix.glob ("*", {MARK=true})`).

Thanks again for working on this.

gvvaughan commented:
Actually, a third option occurs to me: always pass a table of arguments, so that the pattern can still be optional (currently defaults to "*"), and then we can call without parens:

  ```lua
  posix.glob {MARK=true}  -- no pattern
  posix.glob {pattern=".*", MARK=true} -- pattern and flag
  posix.glob (".*") -- for backwards compatibility, equivalent to posix.glob {pattern=".*"}
  ```

telemachus commented:
@gvvaughan This latest version is diagnostic as much as anything. I made a first pass at what you suggested, but I'm not crazy about it.

+ I export `GLOB_MARK`, but I no longer see the point of that since I'm the only one who will ever use it (in `lib/posix.lua.in`). Now that I've changed the function call, it's not possible for anyone else to use the exported constant, which makes me question whether it should be exported at all.
+ I left the failing test trying to diagnose arguments. Since the function can now accept a table, a string, or nothing, I'm not sure how to test that the arguments are good. I'm not even sure that it makes sense to validate the arguments in the Lua-wrapped version since Lua itself tends to be permissive about arguments.
+ Finally, I'm not sure that the wrapper I wrote is what you had in mind. In a nutshell, all I've done is write something to handle all the possible argument variations.

Let me know what you think. If you have any suggestions, I'll give it another pass.

gvvaughan commented on ext/posix/glob.c:
```diff
@@ -29,6 +29,7 @@
 Find all files in this directory matching a shell pattern.
 @function glob
 @string[opt="*"] pat shell glob pattern
+@param[opt="false"] globmark append trailing slash to directories
```
@param flags currently limited to `posix.glob.GLOB_MARK`

gvvaughan commented on ext/posix/glob.c:
```diff
@@ -37,10 +38,11 @@ static int
-	checknargs(L, 1);
-	if (glob(pattern, 0, NULL, &globres))
+	checknargs(L, 2);
+	if (glob(pattern, glob_mark, NULL, &globres))
```
so this would become:

```C
if (glob(pattern, flags, NULL, &globres))
```

gvvaughan commented on lib/posix.lua.in:
```diff
@@ -519,5 +519,27 @@ if _DEBUG ~= false then
   end
 end
 
+local function prepare_args (pat_mark)
```
no need to split this out into a separate function, I think.

gvvaughan commented on lib/posix.lua.in:
```diff
@@ -519,5 +519,27 @@ if _DEBUG ~= false then
+  return pat, mark
+end
+
+local cglob = M.glob
```
I would remove the import of `posix.glob` imports into `M` right at the top of the file, and then no need to cache that here... we can just refer to `posix.glob.glob` and `posix.glob.GLOB_MARK` directly when we need them.

telemachus replied:
I may be confused, but I don't think that works. Unless you mean that I should require `posix` as a whole into the file? That is, if I just run `posix.glob.glob` now, I don't think there's anything there. Maybe I'm misunderstanding what you mean?

gvvaughan commented on lib/posix.lua.in:
```diff
@@ -519,5 +519,27 @@ if _DEBUG ~= false then
+
+M.glob = function (...)
+  local pat, mark = prepare_args(...)
+  return cglob(pat, mark)
```
Here I would save a stack frame and collect and parse the Lua API args inside this one function:

```lua
--- Find all files in this directory matching a shell pattern.
-- Brief description of MARK effects on the result here...
-- @function glob
-- @param[opt={}] string|table args either a glob pattern string,
--    or a table with optional keys `pattern` and `MARK`.
-- @return etc. etc.

M.glob = function (args)
  -- support previous `glob ".*"` style calls for backwards compatibility
  if type(args) == "string" then
    args = {pattern=args}
  elseif type(args) ~= "table" then
    args = {}
  end
  local flags = 0
  if args.MARK then
    flags = posix.glob.GLOB_MARK
  end
  -- in future we can add support for any further glob flags here
  return posix.glob.glob (args.pattern, flags)
end
```

gvvaughan commented on lib/posix.lua.in:
```diff
@@ -519,5 +519,27 @@ if _DEBUG ~= false then
+  local pat, mark = prepare_args(...)
+  return cglob(pat, mark)
+end
 
```
To follow the style of the rest of the functions in this file, we'd then wrap the function above in argument checking if `_DEBUG ~= false`.

telemachus replied:
I tried to do that, but still fail the specs locally. This is as close as I can get:

    - posix file system glob with bad arguments diagnoses argument #1 type not table, string or nil:
      FAILED expectation 1: expecting an error containing any of:
      | bad argument #1 to '?' (table, string or nil expected, got boolean)
      or:
      | bad argument #1 to 'fn' (table, string or nil expected, got boolean)
      but got:
      | ./lib/posix/init.lua:541: bad argument #1 to 'glob' (table or string expected, got boolean)
    
    - posix file system glob with bad arguments diagnoses more than maximum of 1 arguments:
      FAILED expectation 1: expecting an error containing any of:
      | bad argument #2 to 'fn' (no more than 1 argument expected, got 2)
      or:
      | bad argument #2 to '?' (no more than 1 argument expected, got 2)
      but got:
      | ./lib/posix/init.lua:543: bad argument #2 to 'glob' (no more than 1 argument expected, got 2)

Here's the (still intermediate) code where I was trying to figure out how to satisfy `badargsdiagnose`. https://gist.github.com/476d562a266a934ea1a77c1014367743

gvvaughan commented:
Hi,

It looks like you are tickling a bug in Specl here. Please go ahead and push what you have - I'll figure out what I need to do to make Specl happy during the merge.

Cheers,
Gary

> On 10 Oct 2016, at 14:50, Peter Aronoff <notifications@github.com> wrote:
> 
> @telemachus commented on this pull request.
> 
> In lib/posix.lua.in:
> 
> > +    if pat_mark.MARK then
> +      mark = M.GLOB_MARK
> +    end
> +  elseif type (pat_mark) == "string" then
> +    pat = pat_mark
> +  end
> +
> +  return pat, mark
> +end
> +
> +local cglob = M.glob
> +
> +M.glob = function (...)
> +  local pat, mark = prepare_args(...)
> +  return cglob(pat, mark)
> +end
>  
> I tried to do that, but still fail the specs locally. This is as close as I can get:
> 
> - posix file system glob with bad arguments diagnoses argument #1 type not table, string or nil:
>   FAILED expectation 1: expecting an error containing any of:
>   | bad argument #1 to '?' (table, string or nil expected, got boolean)
>   or:
>   | bad argument #1 to 'fn' (table, string or nil expected, got boolean)
>   but got:
>   | ./lib/posix/init.lua:541: bad argument #1 to 'glob' (table or string expected, got boolean)
> 
> - posix file system glob with bad arguments diagnoses more than maximum of 1 arguments:
>   FAILED expectation 1: expecting an error containing any of:
>   | bad argument #2 to 'fn' (no more than 1 argument expected, got 2)
>   or:
>   | bad argument #2 to '?' (no more than 1 argument expected, got 2)
>   but got:
>   | ./lib/posix/init.lua:543: bad argument #2 to 'glob' (no more than 1 argument expected, got 2)
> Here's the (still intermediate) code where I was trying to figure out how to satisfy badargsdiagnose. https://gist.github.com/476d562a266a934ea1a77c1014367743
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

telemachus commented:
Urgh. I just kicked off a review based on a merge-pull on a second machine. There (should be) no actual changes here. Sorry for the noise. I'll squash all this in the end.

telemachus commented:
@gvvaughan Ok, I've just pushed up a second version that does a lot of what you asked for. I know there's a bit more to do, and again, I'll squash all these changes later.

For reference:

+ The Ldoc for the glob function in `lib/posix.lua.in` is a mess. I need to fix that.
+ I'm not sure what you mean by "Extra kudos available if you add specl checks for posix.glob.glob too." Do you mean that even after writing the Lua wrapper, I should be able to reach down and test the underlying C code via specl? If so, I'm not sure how.
+ Also if there's anything that jumps out at you as not in keeping with the overall style of the module, just let me know.

Thanks.

gvvaughan commented on lib/posix.lua.in:
```diff
@@ -519,27 +519,42 @@ if _DEBUG ~= false then
+  M.glob = function (...)
+    local argt = {...}
+    local argtype = type(argt[1])
+    local validtypes = { ["table"] = true, ["string"] = true, ["nil"] = true }
```
same here with scoping... instead of constructing a new validtypes table every time we call, move it into file local scope just after the `if _DEBUG ~= false then` line.

gvvaughan commented on lib/posix.lua.in:
```diff
@@ -519,27 +519,42 @@ if _DEBUG ~= false then
+    local argtype = type(argt[1])
+    local validtypes = { ["table"] = true, ["string"] = true, ["nil"] = true }
+    if validtypes[argtype] ~= true then
+      argtypeerror ("glob", 1, "table, string, or nil", argt[1])
```
and here is the cause of the specl check failure... you're outputting 'table, string, or nil' (with an Oxford comma!  Awesome :) but the spec tests against 'string or table'.  There are versions of the type checking and type error helpers in luaposix, specl and stdlib all written at different times and with their own quirks.  Making this error message and the specl check match will fix things.

For a bit more background, following Lua error message style we don't list implied `nil` of a mismatched argument type in the message:

```lua
> table.unpack({}, {})
stdin:1: bad argument #2 to 'unpack' (number expected, got table)
```

...even though a `nil` argument is perfectly valid, it's not in the diagnostic.

telemachus replied:
I'm not sure that's quite the right diagnosis (though I get you what you mean about not needing to mention `nil`).

Here's what I had to do to squelch the errors locally:

      M.glob = function (...)
        local argt = {...}
        local argtype = type(argt[1])
        if validtypes[argtype] ~= true then
          argtypeerror ("?", 1, "table, string or nil", argt[1]) -- "fn" works too
        elseif #argt > 1 then
          toomanyargerror ("?", 1, #argt) -- "fn" works too
        end
        return glob (...)
      end

1. Remove the Oxford comma (as you noted), but keep the `nil` in the list.
2. Change the function name to either `?` or `fn`, presumably because specl knows that it's an anonymous function. If I keep the name as `"glob"`.

Without the second change, this is what I get:

    - posix file system glob with bad arguments diagnoses argument #1 type not table, string or nil:
      FAILED expectation 1: expecting an error containing any of:
      | bad argument #1 to '?' (table, string or nil expected, got boolean)
      or:
      | bad argument #1 to 'fn' (table, string or nil expected, got boolean)
      but got:
      | ./lib/posix/init.lua:552: bad argument #1 to 'glob' (table, string or nil expected, got boolean)
    
    - posix file system glob with bad arguments diagnoses more than maximum of 1 arguments:
      FAILED expectation 1: expecting an error containing any of:
      | bad argument #2 to 'fn' (no more than 1 argument expected, got 2)
      or:
      | bad argument #2 to '?' (no more than 1 argument expected, got 2)
      but got:
      | ./lib/posix/init.lua:554: bad argument #2 to 'glob' (no more than 1 argument expected, got 2)

For the moment, I've removed the serial comma: that's easy enough. (I don't even have strong feelings about it! I think I do it by habit, frankly.) But I'm leaving the name of the function as `"glob"` since I'm guessing that's a bug in specl (or that I'm doing something larger wrong).

gvvaughan commented:
> On 11 Oct 2016, at 14:01, Peter Aronoff <notifications@github.com> wrote:
> 
> @gvvaughan Ok, I've just pushed up a second version that does a lot of what you asked for.

It does!  Thanks again :)

> I know there's a bit more to do, and again, I'll squash all these changes later.

Either is fine, I tend to use merge commits into master, so squashing on the source branch isn’t a requirement.

> For reference:
> 
> 	• The Ldoc for the glob function in lib/posix.lua.in is a mess. I need to fix that.
> 	• I'm not sure what you mean by "Extra kudos available if you add specl checks for posix.glob.glob too." Do you mean that even after writing the Lua wrapper, I should be able to reach down and test the underlying C code via specl? If so, I'm not sure how.

Yes, luaposix has two APIs… a low level one in the submodules that anyone can use directly; and a higher level API implementing more Luaish interfaces that you get by requiring “posix”.

Specifications for the high level `require “posix”` API are in specs/posix_spec.yaml, such as the posix.glob specs you already started.  Specs for the thin wrapper `require “posix.glob”` API that you are using to implement the high level API would go in specs/posix_glob_spec.yaml.


> 	• Also if there's anything that jumps out at you as not in keeping with the overall style of the module, just let me know.

Nope, looks great!

> Thanks.

Cheers,
Gary

telemachus commented:
Ok. I've made the changes you suggested in the last round (they made good sense to me). I also added new tests and updated the documentation slightly. Along the way, I think I may have discovered a bug in `glob`, and there's something confusing me about specl's behavior.

+ At the moment, if `glob` fails to find any matches, it returns `nil` and an error. Based on my reading of the spec, this sounds wrong. I would have expected it to return a table with nothing in it. I added a test confirming the current behavior, in case I'm wrong about the spec. Either way, it should be documented, right?
+ I wasn't able to get the specs to pass unless I added the new spec file (`posix_glob_spec.yaml`) *last* in the makefile. If I put it in alphabetical order, the specs for the *next* file (`posix_grp_spec.yaml`) would blow up and end the whole run of tests. I may have done something very wrong, but I wanted to let you know.

I'm happy to take another pass at anything you think needs it. Thanks.

telemachus commented on specs/posix_glob_spec.yaml:
```diff
@@ -0,0 +1,81 @@
+      chdir (dir)
+      globlist, errmsg = glob("foo", 0)
+      expect (type(errmsg)).to_be "string"
+      expect (errmsg).to_be("foo:" .. " No such file or directory")
```
By the way, this is odd. I had to glue together these two strings because the test runner wasn't happy with a colon followed by a space, even inside a string. I'm guessing this is an issue with YAML parsing, but I'm not sure.

gvvaughan replied:
yep, YAML treats colon space specially, even if the indentation shows what you wanted :-( You can either use colon space pipe for the description string, or put the whole lua block in single quotes - both of which mark the code as a literal.

gvvaughan commented:
This all looks fine now!

I'll merge and fix up the specs when I'm next at a computer. If tests that follow posix_glob_spec.yaml fail, that likely means something is leaking into the global environment that affects those following test examples.

Thanks so much for working patiently on this!

telemachus commented:
> Thanks so much for working patiently on this!

Thanks for patient shepherding along the way. A number of the components here (e.g., specl, Ldoc) aren't things I use regularly, so I appreciate your guidance.

Any thoughts about whether `glob` should error when there are no matches? (I'm wondering if I misunderstood the C specification.)

gvvaughan commented:
Merged.  Thanks again for your work on this.  I have some more cleanups to do over the next week or three, after which I'll make a new release of luaposix that contains your enhancements.

telemachus commented:
Outstanding! It was fun learning more about how luaposix works. I may look for other ways to contribute if I can.
